### PR TITLE
Extracting package id from .nupkg file

### PR DESCRIPTION
**Pull request by olegmoz**

### CODE REVIEW

**Comment by 0crat** [comment]
This pull request [#4](https://github.com/artipie/nuget-adapter/issues/4) is assigned to @iakunin[/z](https://www.0crat.com/u/iakunin), here is [why](https://www.0crat.com/footprint/CT2E6TK9B/fba20d3c-05e7-4034-ab3f-73347423886c); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @g4s8[/z](https://www.0crat.com/u/g4s8) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer; there will be a monetary reward for this job

File: src/main/java/com/artpie/nuget/PackageBinaryStream.java [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+     *
+     * @return Package metadata parsed from the package.
+     */
+    @SuppressWarnings("PMD.AssignmentInOperand")
+    public PackageMeta readMeta() {
+        try (ZipInputStream zipStream = new ZipInputStream(this.stream)) {
+            ZipEntry entry;
+            while ((entry = zipStream.getNextEntry()) != null) {
+                if (entry.getName().endsWith(".nuspec")) {
+                    final PackageId id = parseNuspecXml(zipStream);
```
Comment (iakunin) [inline src/main/java/com/artpie/nuget/PackageBinaryStream.java]: @olegmoz maybe it's better to get rid of this disposable variable?

File: src/main/java/com/artpie/nuget/PackageBinaryStream.java [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+    /**
+     * Parses .nuspec XML from the stream.
+     *
+     * @param stream Stream to read XML from.
+     * @return Info parsed from the XML.
+     */
+    private static PackageId parseNuspecXml(final InputStream stream) {
+        final Document doc;
+        try {
+            final DocumentBuilderFactory factory = DocumentBuilderFactory.newInstance();
```
Comment (iakunin) [inline src/main/java/com/artpie/nuget/PackageBinaryStream.java]: @olegmoz maybe it's better to get rid of this disposable variable?

File: src/main/java/com/artpie/nuget/PackageBinaryStream.java [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+            final DocumentBuilder builder = factory.newDocumentBuilder();
```
Comment (iakunin) [inline src/main/java/com/artpie/nuget/PackageBinaryStream.java]: @olegmoz maybe it's better to get rid of this disposable variable?

File: src/main/java/com/artpie/nuget/PackageBinaryStream.java [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+            doc = builder.parse(stream);
+        } catch (final IOException | ParserConfigurationException | SAXException ex) {
+            throw new IllegalArgumentException("Failed parsing .nuspec XML", ex);
+        }
+        final XML xml = new XMLDocument(doc);
+        final String id = single(xml, "/package/metadata/id/text()");
```
Comment (iakunin) [inline src/main/java/com/artpie/nuget/PackageBinaryStream.java]: @olegmoz maybe it's better to get rid of this disposable variable?

File: src/main/java/com/artpie/nuget/PackageBinaryStream.java [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+        final String version = single(xml, "/package/metadata/version/text()");
```
Comment (iakunin) [inline src/main/java/com/artpie/nuget/PackageBinaryStream.java]: @olegmoz maybe it's better to get rid of this disposable variable?

File: src/main/java/com/artpie/nuget/PackageBinaryStream.java [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+        }
+    }
+
+    /**
+     * Parses .nuspec XML from the stream.
+     *
+     * @param stream Stream to read XML from.
+     * @return Info parsed from the XML.
+     */
+    private static PackageId parseNuspecXml(final InputStream stream) {
```
Comment (iakunin) [inline src/main/java/com/artpie/nuget/PackageBinaryStream.java]: @olegmoz I think this logic should be extracted to separate class (just leave a puzzle for it). What do you think?

File: src/main/java/com/artpie/nuget/PackageBinaryStream.java [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+     * Exception is thrown if zero or more then 1 values found
+     *
+     * @param xml XML document to read from.
+     * @param xpath XPath expression to select data from the XML.
+     * @return Value found by XPath
+     */
+    private static String single(final XML xml, final String xpath) {
+        final List<String> values = xml.xpath(xpath);
+        if (values.isEmpty()) {
+            final String message = String.format("No values found in path: '%s'", xpath);
```
Comment (iakunin) [inline src/main/java/com/artpie/nuget/PackageBinaryStream.java]: @olegmoz maybe it's better to get rid of this disposable variable?

File: src/main/java/com/artpie/nuget/PackageBinaryStream.java [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+            throw new IllegalArgumentException(message);
+        }
+        if (values.size() > 1) {
+            final String message = String.format("Multiple values found in path: '%s'", xpath);
```
Comment (iakunin) [inline src/main/java/com/artpie/nuget/PackageBinaryStream.java]: @olegmoz maybe it's better to get rid of this disposable variable?

File: src/main/java/com/artpie/nuget/PackageBinaryStream.java [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+
+    /**
+     * Reads single string value from XML by XPath.
+     * Exception is thrown if zero or more then 1 values found
+     *
+     * @param xml XML document to read from.
+     * @param xpath XPath expression to select data from the XML.
+     * @return Value found by XPath
+     */
+    private static String single(final XML xml, final String xpath) {
```
Comment (iakunin) [inline src/main/java/com/artpie/nuget/PackageBinaryStream.java]: @olegmoz you can rewrite this method using cactoos-lib:
```
new UncheckedScalar<>(
    new Ternary<String>(
        values,
        vls -> vls.size() == 1,
        vls -> new ItemAt<>(
            vls,
            0,
            itr -> {
                throw new IllegalArgumentException(
                    new FormattedText(
                        "No values found in path: '%s'",
                        xpath
                    ).asString()
                );
            }
        ).value(),
        vls -> {
            throw new IllegalArgumentException(
                new FormattedText(
                    "Multiple values found in path: '%s'",
                    xpath
                ).asString()
            );
        }
    )
).value();
```

But I don't insist on that.

File: src/test/java/com/artpie/nuget/PackageBinaryStreamTest.java [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+
+/**
+ * Tests for {@link PackageBinaryStream}.
+ *
+ * @since 0.1
+ */
+public class PackageBinaryStreamTest {
+
+    @Test
+    void shouldReadMeta() throws Exception {
```
Comment (iakunin) [inline src/test/java/com/artpie/nuget/PackageBinaryStreamTest.java]: @olegmoz maybe it's better to [cactoos-matchers](use https://github.com/llorllale/cactoos-matchers) here? But again: I don't insist.

File: src/main/java/com/artpie/nuget/PackageBinaryStream.java (hunk 6 above)
Comment (olegmoz) [reply src/main/java/com/artpie/nuget/PackageBinaryStream.java]: I think that is not big enough right now, will extract into separate class if it gets bigger.

File: src/main/java/com/artpie/nuget/PackageBinaryStream.java (hunk 9 above)
Comment (olegmoz) [reply src/main/java/com/artpie/nuget/PackageBinaryStream.java]: Thank you for this example! But I think my code is simpler and I'd prefer to keep it.

File: src/test/java/com/artpie/nuget/PackageBinaryStreamTest.java (hunk 10 above)
Comment (olegmoz) [reply src/test/java/com/artpie/nuget/PackageBinaryStreamTest.java]: I'd prefer to not use cactoos.

File: src/main/java/com/artpie/nuget/PackageBinaryStream.java (hunk 2 above)
Comment (olegmoz) [reply src/main/java/com/artpie/nuget/PackageBinaryStream.java]: Here doc variable is needed to limit try/catch scope, it cannot be removed and there is no reason to do so.

**Comment by olegmoz** [comment]
I do not think that 'disposable variables' are a problem. They usually make the code more obvious by giving names to intermediate computation results.

**Comment by iakunin** [comment]
> I do not think that 'disposable variables' are a problem. They usually make the code more obvious by giving names to intermediate computation results.

@olegmoz I understand your position, however I suggest you to read [this blogpost](https://www.yegor256.com/2015/09/01/redundant-variables-are-evil.html).

@g4s8 just to sum things up:
* we don't enforce using cactoos in artipie
* we don't enforce EO principles in artipie

If it is, than this PR is pretty good to be merged.

@g4s8 by the way, if there are some another review tips for this project I would like to know about them

**Comment by g4s8** [comment]
@iakunin the role of the reviewer is to improve code quality in master branch. You need to report all issues you have managed to find and ask performer to fix them. Performer may disagree, but he need to convince you that you was wrong. Without your approval PR can't be merged, so I'd suggest protecting your comments. Performer has only two possible solutions to resolve review issues: either fix it or explain that reviewer is wrong. Also, you may find this post helpful: https://www.yegor256.com/2015/02/09/serious-code-reviewer.html

File: src/test/java/com/artpie/nuget/PackageBinaryStreamTest.java (hunk 10 above)
Comment (iakunin) [reply src/test/java/com/artpie/nuget/PackageBinaryStreamTest.java]: @olegmoz according to [this comment](https://github.com/artipie/nuget-adapter/pull/4#issuecomment-580614815) I can insist on my point of view.

In my opinion, using [cactoos-matchers](https://github.com/llorllale/cactoos-matchers) will improve code quality: both readability and maintainability.

**Comment by iakunin** [comment]
> I do not think that 'disposable variables' are a problem. They usually make the code more obvious by giving names to intermediate computation results.

@olegmoz according to [this comment](https://github.com/artipie/nuget-adapter/pull/4#issuecomment-580614815) I can insist on my point of view.

I think we should avoid redundant variable as much as possible, because:
> The more variable names I have to remember, the longer it takes to digest the code.

My point of view is described in [this blogpost](https://www.yegor256.com/2015/09/01/redundant-variables-are-evil.html). So, please, get rid of disposable (or redundant) variables as much as possible.

File: src/main/java/com/artpie/nuget/PackageBinaryStream.java (hunk 9 above)
Comment (iakunin) [reply src/main/java/com/artpie/nuget/PackageBinaryStream.java]: @olegmoz this method could be simplified a little bit:
```
final List<String> values = xml.xpath(xpath);
return new UncheckedScalar<>(
    new Ternary<>(
        () -> values.size() == 1,
        new ItemAt<>(values, 0),
        () -> {
            throw new IllegalArgumentException(
                new FormattedText(
                    "Multiple values found in path: '%s'",
                    xpath
                ).asString()
            );
        }
    )
).value();
```

And that's it! =) In my opinion it's much better than chain of multiple if-statements.

File: src/main/java/com/artpie/nuget/PackageBinaryStream.java (hunk 6 above)
Comment (iakunin) [reply src/main/java/com/artpie/nuget/PackageBinaryStream.java]: @olegmoz it's another context of getting `PackageId` from `InputStream` (`PackageBinaryStream#parseNuspecXml`). And this context should be separated from reading zip-stream and filtering `*.nuspec` file (`PackageBinaryStream#readMeta`).

Therefore is should be decoupled into separate class.

File: src/main/java/com/artpie/nuget/PackageBinaryStream.java [hunk 11]
```diff
@@ -0,0 +1,123 @@
...
+    /**
+     * Parses meta information from package content.
+     *
+     * @return Package metadata parsed from the package.
+     */
+    @SuppressWarnings("PMD.AssignmentInOperand")
+    public PackageMeta readMeta() {
+        try (ZipInputStream zipStream = new ZipInputStream(this.stream)) {
+            ZipEntry entry;
+            while ((entry = zipStream.getNextEntry()) != null) {
```
Comment (iakunin) [inline src/main/java/com/artpie/nuget/PackageBinaryStream.java]: @olegmoz we should also handle situation when multiple `*.nuspec` files found.

File: src/main/java/com/artpie/nuget/PackageBinaryStream.java (hunk 2 above)
Comment (iakunin) [reply src/main/java/com/artpie/nuget/PackageBinaryStream.java]: @olegmoz here I mean `factory` variable. `doc` variable cannot be deleted, I agree.

File: src/main/java/com/artpie/nuget/PackageBinaryStream.java (hunk 11 above)
Comment (iakunin) [inline src/main/java/com/artpie/nuget/PackageBinaryStream.java]: @olegmoz also I'd suggest you to create an iterator of ZipEntry. That'll give you and opportunity to use cactoos-decorators such as `Filtered` and `Mapped` in order to solve the same problem.

The main disadvantage of current code is that it's more procedural than object-oriented.

**Comment by olegmoz** [comment]
@iakunin I have decided to change design of this module quite significantly, so I am closing this PR

**Comment by iakunin** [comment]
@olegmoz got it.

@g4s8 should my efforts be paid in this case?

**Comment by 0crat** [comment]
Order was finished: +15 point(s) just awarded to @iakunin[/z](https://www.0crat.com/u/iakunin)

**Comment by 0crat** [comment]
Pull request [#4](https://github.com/artipie/nuget-adapter/issues/4) was not merged, no payment for `ARC`, see [§28](http://www.zerocracy.com/policy.html#28)

**Pull request closed without merging**